### PR TITLE
Fix Dynamic Loading Random Segfaults

### PR DESCRIPTION
**Pull request by permcody**
This PR fixes the dynamic loading issue that we see with applications by removing the over aggressive use of "dlclose". There are also issues with the ExecFlag registration. Specifically, since we are using a Singleton, each application can see the flags of other applications, but this is problematic in the Multiapp case, since the setup of the simulation is split around the setup of Multiapps. The flags "known" to an Application can change due to the Multiapp creating issues during finalizing setup. The "easy" workaround to this is to copy construct the available flags known to an app during the early phase of setting up the App.

Testing: There is no real way to test this since the modules seem to dynamically load reliably already even with the current bug. If this passes module testing, at least we'll know these changes don't disrupt the basic dynamic loading cases in the "misc" module. After this PR is in, we can setup additional application based dynamic loading test cases hopefully in the VTB to make sure this is robust. I have a test case that fails before this patch (Pronghorn loading Bison with both running "simple transient diffusion"). That test case now works on Linux, Intel Macs, and Apple Silicon Macs.

Refs #12266.

### CODE REVIEW

File: framework/include/base/AppFactory.h [hunk 1]
```diff
@@ -115,13 +115,16 @@ class AppFactory
     return _name_to_params_pointer;
   }
 
+  ///@{ Don't allow creation through copy consturction or assignment
```
Comment (lindsayad) [inline framework/include/base/AppFactory.h:118]:
```suggestion
  ///@{ Don't allow creation through copy construction or assignment
```

File: framework/include/base/AppFactory.h [hunk 2]
```diff
@@ -115,13 +115,16 @@ class AppFactory
...
+  AppFactory(AppFactory const &) = delete;
+  void operator=(AppFactory const &) = delete;
```
Comment (lindsayad) [inline framework/include/base/AppFactory.h:120]: What about move construction/assignment?

File: framework/include/base/ExecFlagRegistry.h [hunk 3]
```diff
@@ -9,22 +9,10 @@
...
 #pragma once
 
-#include <shared_mutex>
-
 #include "ExecFlagEnum.h"
 
-namespace moose
-{
-namespace internal
-{
```
Comment (lindsayad) [inline framework/include/base/ExecFlagRegistry.h:19]: I think we could still keep these around so that users are encouraged to use the macros instead of the static method directly

File: framework/include/base/ExecFlagRegistry.h [hunk 4]
```diff
@@ -9,22 +9,10 @@
...
-
 class ExecFlagRegistry;
```
Comment (lindsayad) [inline framework/include/base/ExecFlagRegistry.h:15]:
```suggestion
```

File: framework/include/base/MooseApp.h [hunk 5]
```diff
@@ -1168,8 +1168,14 @@ class MooseApp : public ConsoleStreamInterface,
   /// this map when removing relationship managers/ghosting functors
   std::unordered_map<RelationshipManager *, std::shared_ptr<GhostingFunctor>> _undisp_to_disp_rms;
 
+  struct DynamicLibraryInfo
+  {
+    void * _library_handle;
+    std::unordered_set<std::string> _entry_symbols;
```
Comment (lindsayad) [inline framework/include/base/MooseApp.h:1174]:
```suggestion
    void * library_handle;
    std::unordered_set<std::string> entry_symbols;
```
For public members I like the libMesh paradigm of not having a leading underscore

File: framework/include/base/MooseApp.h [hunk 6]
```diff
@@ -1322,7 +1328,7 @@ class MooseApp : public ConsoleStreamInterface,
   std::shared_ptr<Backup> _cached_backup;
 
   /// Execution flags for this App
-  const ExecFlagEnum & _execute_flags;
+  const ExecFlagEnum _execute_flags;
```
Comment (lindsayad) [inline framework/include/base/MooseApp.h:1331]: Why is this change needed?

File: framework/include/base/Registry.h [hunk 7]
```diff
@@ -222,6 +222,11 @@ class Registry
     return getRegistry()._data_file_paths;
   }
 
+  ///@{ Don't allow creation through copy consturction or assignment
+  Registry(Registry const &) = delete;
+  void operator=(Registry const &) = delete;
```
Comment (lindsayad) [inline framework/include/base/Registry.h:227]: Do you want to do the same for move construction/assignment?

File: framework/include/base/Registry.h [hunk 8]
```diff
@@ -222,6 +222,11 @@ class Registry
     return getRegistry()._data_file_paths;
   }
 
+  ///@{ Don't allow creation through copy consturction or assignment
```
Comment (lindsayad) [inline framework/include/base/Registry.h:225]:
```suggestion
  ///@{ Don't allow creation through copy construction or assignment
```

File: framework/src/base/ExecFlagRegistry.C [hunk 9]
```diff
@@ -8,31 +8,19 @@
...
   return exec_flag_registry;
 }
 
-ExecFlagRegistry::ExecFlagRegistry() {}
-
 const ExecFlagType &
 ExecFlagRegistry::registerFlag(const std::string & name, const bool is_default)
 {
   const auto name_upper = MooseUtils::toUpper(name);
-  std::unique_lock lock(_flags_mutex);
```
Comment (lindsayad) [inline framework/src/base/ExecFlagRegistry.C:34]: Do we no longer have threading concerns?

File: framework/src/base/MooseApp.C [hunk 10]
```diff
@@ -1875,14 +1875,18 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
         break;
       }
     }
-    handle.close();
+    la_handle.close();
   }
 
-  std::string registration_method_name = params.get<std::string>("registration_method");
+  // This should only occur if we have static linkage.
+  if (dl_lib_filename == "")
```
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1882]:
```suggestion
  if (dl_lib_filename.empty())
```

File: framework/src/base/MooseApp.C [hunk 11]
```diff
@@ -1893,63 +1897,59 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
     MooseUtils::checkFileReadable(dl_lib_full_path, false, /*throw_on_unreadable=*/true);
 
 #ifdef LIBMESH_HAVE_DLOPEN
-    void * handle = dlopen(dl_lib_full_path.c_str(), RTLD_LAZY);
+    void * lib_handle = dlopen(dl_lib_full_path.c_str(), RTLD_LAZY);
 #else
-    void * handle = nullptr;
+    void * lib_handle = nullptr;
```
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1902]:
```suggestion
    void * const lib_handle = nullptr;
```

File: framework/src/base/MooseApp.C [hunk 12]
```diff
@@ -1893,63 +1897,59 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
...
-                   "the method should exist in that library.\n");
-#endif
+    auto insert_ret = _lib_handles.insert(std::make_pair(library_filename, lib_info));
+    mooseAssert(insert_ret.second == true, "Error inserting into lib_handles map");
+
+    dyn_lib_it = insert_ret.first;
+  }
 
+  // Library has been loaded, check to see if we've called the requested registration method
+  std::string registration_method = params.get<std::string>("registration_method");
```
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1922]:
```suggestion
  const auto registration_method = params.get<std::string>("registration_method");
```

File: framework/src/base/MooseApp.C [hunk 13]
```diff
@@ -1893,63 +1897,59 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
...
+    // get the pointer to the method in the library.  The dlsym()
+    // function returns a null pointer if the symbol cannot be found,
+    // we also explicitly set the pointer to NULL if dlsym is not
+    // available.
 #ifdef LIBMESH_HAVE_DLOPEN
-      dlclose(handle);
+    void * registration_handle =
+        dlsym(dyn_lib_it->second._library_handle, registration_method.c_str());
+#else
+    void * registration_handle = nullptr;
```
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1935]:
```suggestion
    void * const registration_handle =
        dlsym(dyn_lib_it->second._library_handle, registration_method.c_str());
#else
    void * const registration_handle = nullptr;
```

File: framework/src/base/MooseApp.C [hunk 14]
```diff
@@ -1959,9 +1959,25 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
...
+      entry_sym_from_curr_lib.insert(registration_method);
+    }
+    else
+    {
+
+#ifdef DEBUG
+      // We found a dynamic library that doesn't have a dynamic
+      // registration method in it. This shouldn't be an error, so
+      // we'll just move on.
+
```
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1971]:
```suggestion
```

File: framework/src/base/MooseApp.C [hunk 15]
```diff
@@ -1893,63 +1897,59 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
...
+          register_app_t * reg_ptr = reinterpret_cast<register_app_t *>(&registration_handle);
           (*reg_ptr)();
           break;
         }
         case REGALL:
         {
-          typedef void (*register_app_t)(Factory *, ActionFactory *, Syntax *);
-          register_app_t * reg_ptr = reinterpret_cast<register_app_t *>(&registration_method);
+          using register_app_t = void (*)(Factory *, ActionFactory *, Syntax *);
+          register_app_t * reg_ptr = reinterpret_cast<register_app_t *>(&registration_handle);
```
Comment (lindsayad) [inline framework/src/base/MooseApp.C:1952]:
```suggestion
          register_app_t * const reg_ptr = reinterpret_cast<register_app_t *>(&registration_handle);
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/include/base/MooseApp.h (hunk 6 above)
Comment (permcody) [reply framework/include/base/MooseApp.h:1331]: This is one of the most critical changes in this patch. Without it, an application will see a different set of flags when it's first being setup versus when some of the post "add_multiapp" tasks run. This results in errors about unregistered flags during some of the processing we do in the warehouses to setup objects based on ExecFlags. 

While there may be better designs to this problem, I argue this is the better "workaround" for now. However, it might be prudent to put in a detailed comment about why this is like this. The specific error that I ran into was "Pronghorn" complaining that it didn't know about "XFEM_MARK", an exec flag registered in the "xfem" module used by Bison, but not Pronghorn. I tend to agree that Pronghorn doesn't need storage classes for "XFEM_MARK" and this intentional copy of the flags prevents that.

File: framework/src/base/ExecFlagRegistry.C (hunk 9 above)
Comment (permcody) [reply framework/src/base/ExecFlagRegistry.C:34]: From the Scott Meyer's pattern, the use of a "function static" for the singleton pattern removes that concern. Statics are allocated in the "base" storage (below the heap). There's no possible race condition with this pattern.

**Comment by dschwen** [comment]
Just FYI, we do have the [`DynamicLibraryLoader`](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/utils/DynamicLibraryLoader.C) class, which I added for Abaqus UMAT/UExternalDB support.

**Comment by permcody** [comment]
Thanks @dschwen - I do want to move all of this logic to a separate class. For the purpose of the review though, I made these changes in place. Once we get this in, we should move everything over.

**Comment by dschwen** [comment]
Yeah, just pointing out what we have. The class works well for the plugin loading - no idea if it is a great fit here.

**Comment by permcody** [comment]
It is, just don't want too big of a diff

File: framework/include/base/ExecFlagRegistry.h [hunk 16]
```diff
@@ -49,17 +40,17 @@ class ExecFlagRegistry
...
-  /// So it can be constructed
-  friend ExecFlagRegistry & getExecFlagRegistry();
+  /// Return Singleton instance
+  static ExecFlagRegistry & getExecFlagRegistry();
 
-  /// Mutex for guarding access to _flags
-  std::shared_mutex _flags_mutex;
+  ///@{ Don't allow creation through copy consturction or assignment
+  ExecFlagRegistry(ExecFlagRegistry const &) = delete;
+  void operator=(ExecFlagRegistry const &) = delete;
```
Comment (lindsayad) [inline framework/include/base/ExecFlagRegistry.h:48]: also would like to see the move equivalents deleted here

File: framework/include/base/MooseApp.h (hunk 6 above)
Comment (lindsayad) [reply framework/include/base/MooseApp.h]: Yea I think some of this information would be great as a comment to communicate the intent, otherwise another developer may come through here and change to a reference again, especially as we do not have testing anywhere (yet)

Comment (permcody) [reply framework/include/base/MooseApp.h]: Added

File: framework/include/base/Registry.h (hunk 7 above)
Comment (permcody) [reply framework/include/base/Registry.h]: Yes, but did you notice that the signature was wrong?! It was wrong in the `Registry` class. Fixed everywhere!

File: framework/include/base/ExecFlagRegistry.h (hunk 16 above)
Comment (permcody) [reply framework/include/base/ExecFlagRegistry.h]: Done

File: framework/include/base/Registry.h (hunk 7 above)
Comment (lindsayad) [reply framework/include/base/Registry.h]: I did not until I saw your newest commit 😄

File: framework/include/base/MooseApp.h [hunk 17]
```diff
@@ -1321,8 +1327,14 @@ class MooseApp : public ConsoleStreamInterface,
   /// Cache for a Backup to use for restart / recovery
   std::shared_ptr<Backup> _cached_backup;
 
-  /// Execution flags for this App
-  const ExecFlagEnum & _execute_flags;
+  /**
+   * Execution flags for this App. Note: These are copied on purpose instead of maintaining a
+   * reference to the ExecFlagRegistry registry. In the Multiapp case, the registery may be
```
Comment (lindsayad) [inline framework/include/base/MooseApp.h:1332]:
```suggestion
   * reference to the ExecFlagRegistry registry. In the Multiapp case, the registry may be
```

File: framework/src/base/MooseApp.C [hunk 18]
```diff
@@ -2374,14 +2389,15 @@ MooseApp::attachRelationshipManagers(Moose::RelationshipManagerType rm_type,
...
           // because the DisplacedProblem "nonlinear" DisplacedSystem doesn't have any matrices
-          // for which to do coupling. It's actually horrifying to me that we are adding a coupling
-          // functor, that is going to determine its couplings based on a displaced MeshBase object,
-          // to a System associated with the undisplaced MeshBase object (there is only ever one
-          // EquationSystems object per MeshBase object and visa versa). So here I'm left with the
-          // choice of whether to pass in a MeshBase object that is *not* the MeshBase object that
-          // will actually determine the couplings or to pass in the MeshBase object that is
-          // inconsistent with the System DofMap that we are adding the coupling functor for! Let's
-          // err on the side of *libMesh* consistency and pass properly paired MeshBase-DofMap
+          // for which to do coupling. It's actually horrifying to me that we are adding a
```
Comment (permcody) [inline framework/src/base/MooseApp.C:2392]: Perhaps what's more horrifying is that I'll be credited with owning these lines after a clang-format.

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/include/base/MooseApp.h (hunk 17 above)
Comment (permcody) [reply framework/include/base/MooseApp.h:1332]: F***

File: framework/src/base/MooseApp.C (hunk 18 above)
Comment (lindsayad) [reply framework/src/base/MooseApp.C:2392]: yep everyone will assume you're the guy to come talk to about this. You are the relationship manager after all

Comment (permcody) [reply framework/src/base/MooseApp.C:2392]: 🤣

**Comment by permcody** [comment]
@dschwen - Looking at your DynamicLibraryLoader class now. The problem I ran into with this simpler implementation was library dependencies. I guess none of your UMAT libraries depend on other libraries, but this would fail for those types of cases. In MOOSE we almost always have long dependency chains. I think you could just use my methods directly with one minor change (my methods load AND call a method, looks like yours just loads). Other than that mine has additional functionality. I am looking at moving mine out but should be able to support your `getFunction` method.

**Review by lindsayad — APPROVED** [review]
